_posts/2025-04-25-secure-mcp-sse-server.adoc

```diff
@@ -0,0 +1,399 @@
+
+Next, make a `user-name-provider` tool call:
+
+image::mcp_inspector_tool_call.png[MCP Inspector Tool Call,align="center"]
+
+=== Use curl to access MCP server
+
+Finally, let's use `curl` and also learn a little bit how both the MCP protocol and MCP SSE transport work. 
+
+First, access the main SSE endpoint without the GitHub access token:
+
+[source,shell]
+----
+curl -v localhost:8080/mcp/sse
+----
+
+You will get HTTP 401 error.
+
+Use the access token to access MCP server:
+
+```shell script
+curl -v -H "Authorization: Bearer gho_..." localhost:8080/mcp/sse
+```
+
+and get an SSE response such as:
+
+[source,shell]
+----
+< content-type: text/event-stream
+< 
+event: endpoint
+data: /messages/ZTZjZDE5MzItZDE1ZC00NzBjLTk0ZmYtYThiYTgwNzI1MGJ
+----
+
+The SSE connection is created.
+
+Now open another window and use the same access token to initialize the curl as MCP client, and access the tool, using the value of the `data` property to build the target URL.
+
+Initialize the client:
+
+[source,shell]
+----
+curl -v -H "Authorization: Bearer gho_..." -H "Content-Type: application/json" --data @initialize.json http://localhost:8080/mcp/messages/ZTZjZDE5MzItZDE1ZC00NzBjLTk0ZmYtYThiYTgwNzI1MGJ
+----
+
+where the `initialize.json` file has a content like this:
+
+[source,json]
+----
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "method": "initialize",
+  "params": {
+    "protocolVersion": "2024-11-05",
+    "capabilities": {
+      "roots": {
+        "listChanged": true
+      },
+      "sampling": {}
+    },
+    "clientInfo": {
+      "name": "CurlClient",
+      "version": "1.0.0"
+    }
+  }
+}
+----
+
+and send the initialization notification:
+
+[source,shell]
+----
+curl -v -H "Authorization: Bearer gho_..." -H "Content-Type: application/json" --data @initialized.json http://localhost:8080/mcp/messages/ZTZjZDE5MzItZDE1ZC00NzBjLTk0ZmYtYThiYTgwNzI1MGJ
+----
+
+where the `initialized.json` file has a content like this:
+
+```json
+{
+  "jsonrpc": "2.0",
+  "method": "notifications/initialized"
+}
+```
+
+And call the tool:
+
+[source,shell]
+----
+curl -v -H "Authorization: Bearer gho_..." -H "Content-Type: application/json" --data @call.json http://localhost:8080/mcp/messages/ZTZjZDE5MzItZDE1ZC00NzBjLTk0ZmYtYThiYTgwNzI1MGJ
+----
+
+where the `call.json` file has a content like this:
+
+[source,json]
+----
+{
+  "jsonrpc": "2.0",
+  "id": 2,
+  "method": "tools/call",
+  "params": {
+    "name": "user-name-provider",
+    "arguments": {
+    }
+  }
+}
+----
+
+Now look at the SSE connection window and you will see the name from your GitHub account returned.
+
+== Conclusion
+
+In this blog post, we have explained how you can easily create a secure Quarkus MCP SSE server, obtain an access token and use it to access the MCP server tool in the dev mode with `Quarkus MCP SSE Server Dev UI` and the prod mode with both the https://modelcontextprotocol.io/docs/tools/inspector[MCP inspector] and the curl tools.
+
```

Comment:
maybe add:

"Notice that there is no real difference from adding oauth to MCP or how it is done for a REST endpoint. The most complex part is to get the settings configured for your OpenID provider - but when all is done you just apply a few annotations to mark relevant methods as secure and Quarkus handle the authentication for you."

Reply:
I'll try